### PR TITLE
Add delAllergy command

### LINKED ISSUE #173: Add deleteAllergy command
Improve utility by adding a way to
remove allergy from patient

### PR DESCRIPTION
**Pull request by TechRaj**
- Add a way to delete allergies from patients 
- -Reformatted parts of AllergyCommand parser
to adhere to SLAP and DRY
- Edited MedCon and Allergy related classes to make them more uniform\

Fixes #173

### CODE REVIEW

**Review by hweinian — CHANGES_REQUESTED** [review]

File: src/main/java/seedu/address/logic/commands/DelAllergyCommand.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        requireAllNonNull(nric, allergies);
+
+        this.nric = nric;
+        this.allergies = allergies;
+    }
+
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        Person person = model.fetchPersonIfPresent(new NricMatchesPredicate(nric))
+                .orElseThrow(() -> new CommandException(PATIENT_DOES_NOT_EXIST));
```
[the remark this answers is not in the document]
Comment (TechRaj) [reply src/main/java/seedu/address/logic/commands/DelAllergyCommand.java:53]: Oops did not notice. I will make the necessary amendments

File: src/main/java/seedu/address/logic/parser/AddAllergyCommandParser.java [hunk 2]
```diff
@@ -20,37 +21,36 @@
 public class AddAllergyCommandParser implements Parser<AddAllergyCommand> {
     private static final Logger logger = LogsCenter.getLogger(AddCommandParser.class);
     /**
-     * Parses the given {@code String} of arguments in the context of the AddAllergyCommand
-     * and returns an AddAllergyCommand object for execution.
+     * Parses the given arguments string and creates an {@link AddAllergyCommand} object.
      *
-     * @throws ParseException if the user input does not conform the expected format.
+     * @param args the arguments string containing user input.
+     * @return An {@link AddAllergyCommand} object containing the parsed NRIC and set of allergies.
```
Comment (yong-zhuo) [inline src/main/java/seedu/address/logic/parser/AddAllergyCommandParser.java:27]: In the codebase we use `{@code}` instead of `{@link}`, I am not sure if it violates any coding standards

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 3]
```diff
@@ -205,15 +190,21 @@ public static Appointment parseSerialisedAppointment(String appointment) throws
...
-        final Set<Allergy> allergiesSet = new HashSet<>();
-        for (String allergyName : allergies) {
-            allergiesSet.add(parseAllergy(allergyName));
+    public static Allergy parseAllergy(String allergyStr) throws ParseException {
+        if (allergyStr.isEmpty()) {
+            throw new ParseException("Allergy " + MESSAGE_EMPTY_FIELD);
+        } else if (allergyStr.length() > 30) {
+            throw new ParseException("Allergy " + MESSAGE_CONSTRAINTS_LENGTH);
+        } else if (!allergyStr.matches(Allergy.VALIDATION_REGEX)) {
+            throw new ParseException("Invalid allergy format.");
```
[the remark this answers is not in the document]
Comment (TechRaj) [reply src/main/java/seedu/address/logic/parser/ParserUtil.java:205]: Nope there isn't. I just redid an existing method in parseUtil to reduce code in the main execute function

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 4]
```diff
@@ -205,15 +190,21 @@ public static Appointment parseSerialisedAppointment(String appointment) throws
...
+     *
+     * @param allergyStr The medical condition string to parse.
+     * @return A {@code Allergy} object.
+     * @throws ParseException If the allergy string is invalid.
      */
-    public static Set<Allergy> parseAllergies(Collection<String> allergies) throws ParseException {
-        requireNonNull(allergies);
-        final Set<Allergy> allergiesSet = new HashSet<>();
-        for (String allergyName : allergies) {
-            allergiesSet.add(parseAllergy(allergyName));
```
Comment (yong-zhuo) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:214]: This was how the original `parseTags` was done, I am not sure how your new `parseAllergy` will affect the command

File: src/main/java/seedu/address/model/person/Allergy.java [hunk 5]
```diff
@@ -22,15 +24,19 @@ public class Allergy implements Comparable<Allergy> {
      */
     public Allergy(String allergyName) {
         requireNonNull(allergyName);
-        checkArgument(isValidAllergyName(allergyName), MESSAGE_CONSTRAINTS);
+        checkArgument(!allergyName.isEmpty(), MESSAGE_EMPTY_FIELD);
+        checkArgument(isValidAllergyName(allergyName), MESSAGE_CONSTRAINTS_LENGTH);
```
[the remark this answers is not in the document]
Comment (TechRaj) [reply src/main/java/seedu/address/model/person/Allergy.java:28]: I could remedy it with adding another checkArgument specifically for checking alpha numerics, what are your thoughts?

File: src/main/java/seedu/address/logic/commands/DelAllergyCommand.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    public static final String PATIENT_DOES_NOT_EXIST = "Patient does not exist in contact list";
+    public static final String MESSAGE_ALLERGY_NOT_FOUND = "Allergy not found: %1$s";
+
+    private final Nric nric;
+    private final Set<Allergy> allergies;
+
+    /**
+     * @param nric of the patient to unassign the allergies from.
+     * @param allergies set of allergies to be removed.
+     */
```
[the remark this answers is not in the document]
Comment (TechRaj) [reply src/main/java/seedu/address/logic/commands/DelAllergyCommand.java:42]: noted. thanks for the catch!

File: src/main/java/seedu/address/logic/commands/DelAllergyCommand.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            model.setPerson(person, editedPerson);
+            model.updateFilteredPersonList(Model.PREDICATE_SHOW_ALL_PERSONS);
+            return new CommandResult(generateSuccessMessage(editedPerson));
+        }
+        throw new CommandException(PATIENT_DOES_NOT_EXIST);
+    }
+
+    /**
+     * Generates a command execution success message based on the removed medical conditions.
+     */
```
[the remark this answers is not in the document]
Comment (TechRaj) [reply src/main/java/seedu/address/logic/commands/DelAllergyCommand.java:81]: noted

File: src/main/java/seedu/address/logic/commands/DelAllergyCommand.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+            // create an edited person with the updated allergies
+            Person editedPerson = new Person(
+                    person.getName(), person.getPhone(), person.getEmail(),
+                    person.getNric(), person.getAddress(), person.getDateOfBirth(),
+                    person.getGender(), updatedAllergySet, person.getPriority(), person.getAppointments(),
+                    person.getMedCons());
+
+            model.setPerson(person, editedPerson);
+            model.updateFilteredPersonList(Model.PREDICATE_SHOW_ALL_PERSONS);
+            return new CommandResult(generateSuccessMessage(editedPerson));
```
Comment (PresleyChew) [inline src/main/java/seedu/address/logic/commands/DelAllergyCommand.java:77]: LGTM

File: src/main/java/seedu/address/logic/parser/DelAllergyCommandParser.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+     * @param args the arguments string containing user input.
+     * @return A {@link DelAllergyCommand} object containing the parsed NRIC and set of allergies.
+     * @throws ParseException if the user input does not conform to the expected format or
+     *         if the NRIC is not provided.
+     */
+    public DelAllergyCommand parse(String args) throws ParseException {
+        requireNonNull(args);
+        ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args, PREFIX_NRIC, PREFIX_ALLERGY);
+
+        // Check if NRIC is provided
```
Comment (PresleyChew) [inline src/main/java/seedu/address/logic/parser/DelAllergyCommandParser.java:37]: Change to 
`Check if NRIC and Allergy is provided `
since PREFIX_NRIC PREFIX_ALLERGY in parameter for checking of prefix.

File: src/main/java/seedu/address/logic/parser/DelAllergyCommandParser.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        if (!argMultimap.arePrefixesPresent(PREFIX_NRIC, PREFIX_ALLERGY) || !argMultimap.getPreamble().isEmpty()) {
+            logger.warning("NRIC not provided in DelAllergyCommand arguments.");
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, DelAllergyCommand.MESSAGE_USAGE));
+        }
+
+        argMultimap.verifyNoDuplicatePrefixesFor(PREFIX_NRIC);
```
Comment (PresleyChew) [inline src/main/java/seedu/address/logic/parser/DelAllergyCommandParser.java:43]: Good check to ensure no duplicate NRIC parameter

File: src/main/java/seedu/address/logic/parser/DelAllergyCommandParser.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+        // Parse NRIC
+        String nricStr = argMultimap.getValue(PREFIX_NRIC).get();
+        Nric nric = ParserUtil.parseNric(nricStr);
+
+        // Parse all Allergy values and add them to a set
+        Set<Allergy> allergies = new HashSet<>();
+        for (String allergyStr : argMultimap.getAllValues(PREFIX_ALLERGY)) {
+            Allergy allergy = ParserUtil.parseAllergy(allergyStr);
+            allergies.add(allergy);
+        }
```
Comment (PresleyChew) [inline src/main/java/seedu/address/logic/parser/DelAllergyCommandParser.java:54]: Good use of HashSet!

File: src/main/java/seedu/address/model/person/MedCon.java [hunk 12]
```diff
@@ -12,18 +12,18 @@
 public class MedCon implements Comparable<MedCon> {
 
     public static final String VALIDATION_REGEX = "\\p{Alnum}+(\\s\\p{Alnum}+)*";
-    public final String value;
+    public final String medConName;
```
Comment (PresleyChew) [inline src/main/java/seedu/address/model/person/MedCon.java:15]: Good change of name to better fix the intended purpose!

**Review by PresleyChew — COMMENTED** [review]
Overall LGTM, after amending what @hweinian and @yong-zhuo proposed, as well as the change in the comment for what I have commented, should be ready to merge.

File: src/main/java/seedu/address/logic/parser/DelAllergyCommandParser.java (hunk 9 above)
Comment (TechRaj) [reply src/main/java/seedu/address/logic/parser/DelAllergyCommandParser.java]: Done. Good catch!

File: src/main/java/seedu/address/logic/parser/DelAllergyCommandParser.java (hunk 11 above)
Comment (TechRaj) [reply src/main/java/seedu/address/logic/parser/DelAllergyCommandParser.java:54]: It has always used a hashset. But thanks!

File: src/main/java/seedu/address/logic/parser/DelAllergyCommandParser.java (hunk 10 above)
Comment (TechRaj) [reply src/main/java/seedu/address/logic/parser/DelAllergyCommandParser.java:43]: yep

File: src/main/java/seedu/address/model/person/MedCon.java (hunk 12 above)
Comment (TechRaj) [reply src/main/java/seedu/address/model/person/MedCon.java:15]: yep

File: src/main/java/seedu/address/logic/commands/DelAllergyCommand.java [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        this.allergies = allergies;
+    }
+
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        Person person = model.fetchPersonIfPresent(new NricMatchesPredicate(nric))
+                .orElseThrow(() -> new CommandException(MESSAGE_PERSON_NRIC_NOT_FOUND));
+
+        if (person.getNric().equals(this.nric)) {
+            Set<Allergy> updatedAllergySet = new HashSet<>(person.getAllergies());
```
Comment (yadobler) [inline src/main/java/seedu/address/logic/commands/DelAllergyCommand.java:59]: Good use of guard clause

use this for `AddAllergyCommand.java`

File: src/main/java/seedu/address/logic/commands/DelAllergyCommand.java [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+                    person.getName(), person.getPhone(), person.getEmail(),
+                    person.getNric(), person.getAddress(), person.getDateOfBirth(),
+                    person.getGender(), updatedAllergySet, person.getPriority(), person.getAppointments(),
+                    person.getMedCons());
+
+            model.setPerson(person, editedPerson);
+            model.updateFilteredPersonList(Model.PREDICATE_SHOW_ALL_PERSONS);
+            return new CommandResult(generateSuccessMessage(editedPerson));
+        }
+        throw new CommandException(PATIENT_DOES_NOT_EXIST);
```
Comment (yadobler) [inline src/main/java/seedu/address/logic/commands/DelAllergyCommand.java:79]: This is redundant since the guarding predicate should catch the patient not existing - it is also why the coverage fails, because it is unreachable code

File: src/main/java/seedu/address/logic/commands/DelMedConCommand.java
Comment (yadobler) [inline src/main/java/seedu/address/logic/commands/DelMedConCommand.java:1]: This should be in another PR with the other `MedCon` edits

File: src/main/java/seedu/address/logic/parser/AddAllergyCommandParser.java [hunk 15]
```diff
@@ -20,37 +21,36 @@
...
-        ArgumentMultimap argMultimap =
-                ArgumentTokenizer.tokenize(args, PREFIX_NRIC, PREFIX_ALLERGY);
+        requireNonNull(args);
+        ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args, PREFIX_NRIC, PREFIX_ALLERGY);
 
-        if (!arePrefixesPresent(argMultimap, PREFIX_NRIC, PREFIX_ALLERGY)
-                || !argMultimap.getPreamble().isEmpty()) {
+        // Check if NRIC and allergy is provided
+        if (!argMultimap.arePrefixesPresent(PREFIX_NRIC, PREFIX_ALLERGY) || !argMultimap.getPreamble().isEmpty()) {
+            logger.warning("NRIC and allergy not provided in AddAllergyCommand arguments.");
```
Comment (yadobler) [inline src/main/java/seedu/address/logic/parser/AddAllergyCommandParser.java:37]: If preamble is not empty, then this warning is not correct. I think can just append to this warning text that unexpected preamble was detected

File: src/main/java/seedu/address/logic/parser/AddressBookParser.java [hunk 16]
```diff
@@ -110,6 +111,9 @@ public Command parseCommand(String userInput) throws ParseException {
         case AddAllergyCommand.COMMAND_WORD:
             return new AddAllergyCommandParser().parse(arguments);
 
+        case DelAllergyCommand.COMMAND_WORD:
+            return new DelAllergyCommandParser().parse(arguments);
```
Comment (yadobler) [inline src/main/java/seedu/address/logic/parser/AddressBookParser.java:115]: add to `AddressBookParserTest`:

```java
    @Test
    public void parseCommand_delAllergy() throws Exception {
        String commandInput = DelAllergyCommand.COMMAND_WORD + " " + NRIC_DESC_AMY
                + " " + PREFIX_ALLERGY + "Insulin";
        DelAllergyCommand command = (DelAllergyCommand) parser.parseCommand(commandInput);
        Set<Allergy> allergies = new HashSet<Allergy>();
        allergies.add(new Allergy("Insulin"));
        assertEquals(new DelAllergyCommand(new Nric(VALID_NRIC_AMY), allergies), command);
    }
```

File: src/main/java/seedu/address/logic/parser/EditCommandParser.java [hunk 17]
```diff
@@ -73,30 +67,11 @@ public EditCommand parse(String args) throws ParseException {
...
-        assert allergies != null;
-
-        if (allergies.isEmpty()) {
-            return Optional.empty();
-        }
-        Collection<String> allergiesSet = allergies.size() == 1 && allergies.contains("")
-                ? Collections.emptySet()
-                : allergies;
-        return Optional.of(ParserUtil.parseAllergies(allergiesSet));
-    }
```
Comment (yadobler) [inline src/main/java/seedu/address/logic/parser/EditCommandParser.java:101]: Good point deleting this since `Edit` should not be able to touch allergies

File: src/test/java/seedu/address/model/person/AllergyTest.java [hunk 18]
```diff
@@ -17,10 +17,4 @@ public void constructor_invalidTagName_throwsIllegalArgumentException() {
         assertThrows(IllegalArgumentException.class, () -> new Allergy(invalidTagName));
     }
 
-    @Test
-    public void isValidTagName() {
-        // null tag name
-        assertThrows(NullPointerException.class, () -> Allergy.isValidAllergyName(null));
-    }
-
```
Comment (yadobler) [inline src/test/java/seedu/address/model/person/AllergyTest.java:25]: You need to test `isValidAllergyName` separately because it is a `public static` method that can be used without the `Allergy()` constructor.

File: src/main/java/seedu/address/model/person/MedCon.java [hunk 19]
```diff
@@ -12,18 +12,18 @@
...
      */
-    public MedCon(String medCon) {
-        requireNonNull(medCon);
-        checkArgument(!medCon.isEmpty(), MESSAGE_EMPTY_FIELD);
-        checkArgument(isValidMedCon(medCon), MESSAGE_CONSTRAINTS_LENGTH);
-        value = medCon;
+    public MedCon(String medConName) {
+        requireNonNull(medConName);
+        checkArgument(!medConName.isEmpty(), MESSAGE_EMPTY_FIELD);
+        checkArgument(isValidMedConName(medConName), MESSAGE_CONSTRAINTS_ALPHANUMERIC_LENGTH);
```
Comment (yadobler) [inline src/main/java/seedu/address/model/person/MedCon.java:25]: Good use of multiple error messages to provide users more info

Comment (yadobler) [reply src/main/java/seedu/address/model/person/MedCon.java:25]: But might be violating `DRY` since `isValidMedConName` is already checking nullity. Why not let `isValidMedConName` check and throw the Exceptions (ie push the `checkArgument` calls into `isValidMedConName`), and let the Exception propagate up and out of the constructor via `isValidMedConName` instead. It will reduce coverage misses too.

(if you choose this route, rename the function)

(take a look at `Appointment::checkTimePeriodValid`)

File: src/main/java/seedu/address/logic/commands/AddAllergyCommand.java [hunk 20]
```diff
@@ -54,39 +55,37 @@ public CommandResult execute(Model model) throws CommandException {
...
-        if (person.getNric().equals(this.nric)) {
-            Set<Allergy> updatedAllergySet = new HashSet<>(person.getAllergies());
-            // Check for duplicates
-            for (Allergy allergy : allergies) {
-                if (!updatedAllergySet.add(allergy)) {
-                    throw new CommandException(String.format(MESSAGE_DUPLICATE_ALLERGY, allergy.allergyName));
-                }
+        // if the NRICs do not match, throw exception immediately.
+        if (!person.getNric().equals(this.nric)) {
+            throw new CommandException(PATIENT_DOES_NOT_EXIST);
```
Comment (yadobler) [inline src/main/java/seedu/address/logic/commands/AddAllergyCommand.java:60]: good guard clause

File: src/main/java/seedu/address/logic/commands/AddAllergyCommand.java [hunk 21]
```diff
@@ -54,39 +55,37 @@ public CommandResult execute(Model model) throws CommandException {
...
-
-        // Remove trailing comma and space, if any
-        if (allergiesString.length() > 0) {
-            allergiesString.setLength(allergiesString.length() - 2);
-        }
-
-        String resultAllergies = '[' + allergiesString.toString() + ']';
+        String resultAllergies = allergies.stream()
+                .map(allergy -> allergy.allergyName)
+                .collect(Collectors.joining(", ", "[", "]"));
```
Comment (yadobler) [inline src/main/java/seedu/address/logic/commands/AddAllergyCommand.java:88]: Nice use of streams

File: src/main/java/seedu/address/logic/parser/AddAllergyCommandParser.java [hunk 22]
```diff
@@ -34,7 +34,7 @@ public AddAllergyCommand parse(String args) throws ParseException {
 
         // Check if NRIC and allergy is provided
         if (!argMultimap.arePrefixesPresent(PREFIX_NRIC, PREFIX_ALLERGY) || !argMultimap.getPreamble().isEmpty()) {
-            logger.warning("NRIC and allergy not provided in AddAllergyCommand arguments.");
+            logger.warning("Unexpected preamble detected.");
```
Comment (yadobler) [inline src/main/java/seedu/address/logic/parser/AddAllergyCommandParser.java:37]: I mean it could also be missing `NRIC` and `Allergy` prefixes but sure i guess

File: src/test/java/seedu/address/logic/commands/AddAllergyCommandTest.java [hunk 23]
```diff
@@ -59,6 +62,22 @@ public void execute_addAllergy_success() throws Exception {
...
+
+        Set<Allergy> duplicateAllergies = new HashSet<>(ALICE.getAllergies());
+        AddAllergyCommand command = new AddAllergyCommand(nric, duplicateAllergies);
+
+        Allergy firstAllergy = ALICE.getAllergies().iterator().next();
+        String expectedMessage = String.format(MESSAGE_DUPLICATE_ALLERGY, firstAllergy.allergyName);
+
+        assertCommandFailure(command, model, expectedMessage);
+    }
+
```
Comment (yadobler) [inline src/test/java/seedu/address/logic/commands/AddAllergyCommandTest.java:79]: good test

**Review by yadobler — APPROVED** [review]
LGTM

File: src/main/java/seedu/address/model/person/MedCon.java (hunk 19 above)
Comment (yadobler) [reply src/main/java/seedu/address/model/person/MedCon.java:25]: If you wanna quickly make an _issue_ for the PR, just press the 3 dots and `Reference in new issue`
